### PR TITLE
CA-214 (3/3): feat: build FavoritesSection widget with l10n and tests

### PR DESCRIPTION
**Pull request by ayanasamuel8**
## PR SUMMARY

Third of 3 stacked PRs for CA-214 (DASH-011). Adds the `FavoritesSection` widget and required l10n keys:

- `FavoritesSection` — stateless widget that composes `FavoriteCalculationCard` and `FavoriteEstimationCard` in a vertical `Column` with a section header ("Favourites" + "View all" inline `TextButton`) and an empty-state `Text`; uses `for` loops with conditional `SizedBox` separators (avoids index arithmetic per Rule 12)
- `app_en.arb` — adds `favouritesSectionTitle` ("Favourites") and `favouritesEmptyState` ("No favourites yet.") l10n keys

All spacing, colors, and typography use CoreUI tokens. No loading or error state (blocked on CA-247 data-layer ticket).

**Stacking order:** [1/3 Entities + FavoriteCalculationCard #388] → [2/3 FavoriteEstimationCard #389] → This PR
**Base PR:** #389

## REVIEW SUMMARY

| Issue Name | Description | Status | Notes |
|------------|-------------|--------|-------|
| Rule 8 — find.byType for card widgets in section test | Tests use `find.byType(FavoriteCalculationCard)` and `find.byType(FavoriteEstimationCard)` to assert card count and presence | Disagree | These are first-party widget types owned by this feature, not implementation-detail CoreUI internals. The pattern is consistent with existing project tests (e.g. `find.byType(EstimationCard)` in `recent_estimations_section_test.dart`). Using `byType` here is appropriate since the section's contract is explicitly "render one card per item". |
| Rule 14 — No a11y test | New composable section widget with interactive children introduced without a11y coverage | ✅ Fixed | Added `test/features/dashboard/widgets/accessibility/favorites_section_a11y_test.dart`; tests empty and loaded states in both light and dark themes. Tap target size opted out: View All shrink-wrap button and small chips are prescribed by the Figma spec. Text contrast opted out for loaded state: chip label tokens achieve ~3.73:1 contrast, a CoreUI design token gap. |

## Test plan

- [x] `fvm flutter test test/features/dashboard/widgets/favorites_section_test.dart` — 9 tests pass
- [x] `fvm flutter test test/features/dashboard/screenshots/favorites_section_screenshot_test.dart` — 2 goldens pass
- [x] `fvm flutter test test/features/dashboard/widgets/accessibility/favorites_section_a11y_test.dart` — 2 tests pass
- [x] `fvm dart run custom_lint` — no issues

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: test/features/dashboard/widgets/favorites_section_test.dart [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        ),
+      ),
+    );
+    await tester.pumpAndSettle();
+  }
+
+  testWidgets('renders section title', (tester) async {
+    await pumpSection(tester);
+    expect(find.text('Favourites'), findsOneWidget);
+  });
```
Comment (ripplearcgit) [inline test/features/dashboard/widgets/favorites_section_test.dart:71]: **Nit: hardcoded l10n strings** — the project convention is `lookupAppLocalizations(const Locale('en'))` for all text assertions so that ARB copy-editing changes don't silently break tests (see `recent_estimations_section_test.dart`).

```dart
// Add at top of main():
final l10n = lookupAppLocalizations(const Locale('en'));

// Replace hardcoded strings:
expect(find.text(l10n.favouritesSectionTitle), findsOneWidget);  // line 63
expect(find.text(l10n.viewAllButton), findsOneWidget);           // line 68
expect(find.text(l10n.favouritesEmptyState), findsOneWidget);    // line 73
await tester.tap(find.text(l10n.viewAllButton));                 // line 126
```

File: test/features/dashboard/widgets/accessibility/favorites_section_a11y_test.dart [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+import '../../../../utils/a11y/a11y_guidelines.dart';
+import '../../../../utils/screenshot/font_loader.dart';
+
+void main() {
+  setUpAll(() async {
+    await loadAppFontsAll();
+  });
+
+  final _calculations = [
```
Comment (ripplearcgit) [inline test/features/dashboard/widgets/accessibility/favorites_section_a11y_test.dart]: **Nit: misleading underscore prefix** — `_calculations` and `_estimations` are local variables inside `main()`, not library-private class members. The leading underscore is a Dart convention for library-private scope and the `private_subject` custom lint rule may flag these. Rename to `calculations` and `estimations` to match the pattern in `favorites_section_test.dart`.

File: lib/features/dashboard/presentation/widgets/favorites_section.dart [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import 'package:construculator/features/dashboard/domain/entities/favorite_calculation_entity.dart';
+import 'package:construculator/features/dashboard/domain/entities/favorite_estimation_entity.dart';
+import 'package:construculator/features/dashboard/presentation/widgets/favorite_calculation_card.dart';
+import 'package:construculator/features/dashboard/presentation/widgets/favorite_estimation_card.dart';
+import 'package:construculator/libraries/extensions/extensions.dart';
+import 'package:flutter/material.dart';
+import 'package:ripplearc_coreui/ripplearc_coreui.dart';
+
+/// Displays the Favourites section on the dashboard, combining favorited
+/// calculations and cost estimations in a vertical list.
```
Comment (ripplearcgit) [inline lib/features/dashboard/presentation/widgets/favorites_section.dart]: **Nit: incomplete class-level doc** — The two-line `///` comment says "combining favorited calculations and cost estimations in a vertical list" but omits the empty-state behaviour and the "View all" callback. A reader relying on the doc would not know this widget renders a header row with a `View all` button or that it has a distinct empty state. One sentence update is enough:

```dart
/// Displays the Favourites section on the dashboard. Shows a header row with
/// a "View all" button and, when [calculations] and [estimations] are both
/// empty, renders a centred empty-state message; otherwise renders a vertical
/// list of [FavoriteCalculationCard]s followed by [FavoriteEstimationCard]s.
```

File: lib/features/dashboard/presentation/widgets/favorites_section.dart [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+class FavoritesSection extends StatelessWidget {
+  final List<FavoriteCalculation> calculations;
+  final List<FavoriteEstimation> estimations;
```
Comment (ripplearcgit) [inline lib/features/dashboard/presentation/widgets/favorites_section.dart:20]: **Nit: public fields need `///` doc comments** — All five public parameters (`calculations`, `estimations`, `onViewAll`, `onCalculationTap`, `onEstimationTap`) are undocumented. Project conventions require `///` doc on every public member. A one-liner each is sufficient, e.g.:

```dart
/// The list of favourited calculations to display.
final List<FavoriteCalculation> calculations;

/// The list of favourited estimations to display.
final List<FavoriteEstimation> estimations;

/// Called when the "View all" button is tapped.
final VoidCallback onViewAll;

/// Called with the [FavoriteCalculation.id] when a calculation card is tapped.
final void Function(String id) onCalculationTap;

/// Called with the [FavoriteEstimation.id] when an estimation card is tapped.
final void Function(String id) onEstimationTap;
```

File: lib/features/dashboard/presentation/widgets/favorites_section.dart [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        ),
+        const SizedBox(height: CoreSpacing.space4),
+        if (isEmpty)
+          Center(
+            child: Text(
+              context.l10n.favouritesEmptyState,
+              style: typography.bodyMediumRegular.copyWith(
+                color: colors.textBody,
+              ),
+            ),
```
Comment (ripplearcgit) [inline lib/features/dashboard/presentation/widgets/favorites_section.dart:116]: **Nit (non-blocking): separator index arithmetic** — The condition `i < calculations.length - 1 || estimations.isNotEmpty` is correct (all four boundary cases verified), but it's harder to read at a glance than the PR description implied. Consider extracting a small helper that makes the intent explicit:

```dart
List<Widget> _withSeparators(List<Widget> items) {
  if (items.isEmpty) return items;
  return [
    for (int i = 0; i < items.length; i++) ...[
      items[i],
      if (i < items.length - 1) const SizedBox(height: CoreSpacing.space3),
    ],
  ];
}

// In build():
..._withSeparators([
  ..._withSeparators([for (final c in calculations) FavoriteCalculationCard(...)]),
  ..._withSeparators([for (final e in estimations) FavoriteEstimationCard(...)]),
]),
```

Alternatively, concatenate both lists into one before the loop. Either approach removes the cross-list boundary condition from `build()`. Not blocking.

File: test/features/dashboard/screenshots/favorites_section_screenshot_test.dart [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import 'package:construculator/features/dashboard/presentation/widgets/favorites_section.dart';
+import 'package:construculator/l10n/generated/app_localizations.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import '../../../utils/screenshot/font_loader.dart';
+
+void main() {
+  const size = Size(390, 500);
+  const ratio = 1.0;
```
Comment (ripplearcgit) [inline test/features/dashboard/screenshots/favorites_section_screenshot_test.dart:14]: **Nit: redundant `ensureInitialized()` call** — `TestWidgetsFlutterBinding.ensureInitialized()` is called here at the top of `main()` but `loadAppFontsAll()` already calls it internally (confirmed in `test/utils/screenshot/font_loader.dart`). The standalone call is harmless but adds noise. Compare peer screenshot tests — most omit this top-level call and rely on `loadAppFontsAll()` inside `setUp`.

**Review by ripplearcgit — APPROVED** [review]
## ⚠ Conditional LGTM — CA-214 (3/3): FavoritesSection

The widget itself is architecturally clean — pure stateless compositor, CoreUI tokens only, callbacks for every interaction, correct empty-state guard, and full test coverage (widget + golden + a11y). The l10n files are in sync and the British-English spelling is consistent throughout the stack.

**Blocking:** None.

**Nits to address before merge:**

1. **`favorites_section_test.dart`** — Hardcoded `find.text('Favourites')`, `find.text('View all')`, `find.text('No favourites yet.')` calls should use `lookupAppLocalizations(const Locale('en'))` per the project's existing pattern (see `recent_estimations_section_test.dart`). A copy-editing change to the ARB would silently break these tests.
2. **`favorites_section_a11y_test.dart`** — Top-level `_calculations` / `_estimations` variables inside `main()` have a misleading underscore prefix (Dart convention for library-private, not local scope). Rename to `calculations` / `estimations`.
3. **`favorites_section.dart`** — Five public fields (`calculations`, `estimations`, `onViewAll`, `onCalculationTap`, `onEstimationTap`) lack `///` doc comments. Class-level doc also omits the empty-state and "View all" callback from its description.
4. **`favorites_section_screenshot_test.dart`** — Redundant `TestWidgetsFlutterBinding.ensureInitialized()` at the top of `main()` — `loadAppFontsAll()` already calls it internally.
5. **`favorites_section.dart` lines 69 & 74** — Index arithmetic in the separator condition (`i < calculations.length - 1 || estimations.isNotEmpty`) is correct but adds cognitive overhead. Consider extracting a `_withSeparators` helper (non-blocking, see inline comment).

> Full review report: `~/Desktop/pr-reviews/pr-review-construculator-app-390.html`

File: test/features/dashboard/widgets/favorites_section_test.dart [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        theme: createTestTheme(),
+        locale: const Locale('en'),
+        localizationsDelegates: AppLocalizations.localizationsDelegates,
+        supportedLocales: AppLocalizations.supportedLocales,
+        home: Scaffold(
+          body: FavoritesSection(
+            calculations: calculations,
+            estimations: estimations,
+            onViewAll: onViewAll ?? () {},
+            onCalculationTap: onCalculationTap ?? (_) {},
```
Comment (ripplearcgit) [inline test/features/dashboard/widgets/favorites_section_test.dart:57]: **Nit: hardcoded l10n strings** — project standard is `lookupAppLocalizations(const Locale('en'))` (see `recent_estimations_section_test.dart`), not hardcoded strings, so ARB copy-edits don't silently break tests. Replace:

```dart
// Add at top of main():
final l10n = lookupAppLocalizations(const Locale('en'));

expect(find.text(l10n.favouritesSectionTitle), findsOneWidget);   // line 63
expect(find.text(l10n.viewAllButton), findsOneWidget);            // line 68
expect(find.text(l10n.favouritesEmptyState), findsOneWidget);     // line 73
await tester.tap(find.text(l10n.viewAllButton));                  // line 126
```

**Review by ripplearcgit — APPROVED** [review]
## Conditional LGTM — CA-214 (3/3): FavoritesSection widget

The composition approach is architecturally sound. `FavoritesSection` is a pure stateless compositor: no state, no BLoC dependency, CoreUI tokens throughout, callbacks for all interactions. The l10n wiring is clean (ARB + generated files in sync, British English spelling consistent, `viewAllButton` correctly re-used from #389). Golden tests and a11y tests are thorough and structured correctly.

A few nits to address before or alongside merge:

---

### Nits (non-blocking, but please fix)

**1. Widget test — hardcoded l10n strings** (`favorites_section_test.dart` lines 63, 68, 73, 126)
The project standard (see `recent_estimations_section_test.dart`) is `lookupAppLocalizations(const Locale('en'))` — never hardcoded strings — so ARB copy-edits don't silently break tests. The test harness already sets up `AppLocalizations.localizationsDelegates`, so the lookup is free:
```dart
// Add at top of main():
final l10n = lookupAppLocalizations(const Locale('en'));

// Replace:
expect(find.text(l10n.favouritesSectionTitle), findsOneWidget);   // line 63
expect(find.text(l10n.viewAllButton), findsOneWidget);            // line 68
expect(find.text(l10n.favouritesEmptyState), findsOneWidget);     // line 73
await tester.tap(find.text(l10n.viewAllButton));                  // line 126
```

**2. A11y test — misleading underscore prefix** (`favorites_section_a11y_test.dart` lines 15 & 19)
`_calculations` and `_estimations` are local variables inside `main()`, not class members. The `_` prefix is a Dart library-private convention; applied to locals it may trigger `private_subject` lint and confuses readers. Rename to `calculations` / `estimations` (no underscore), matching the pattern in `favorites_section_test.dart`.

**3. Widget — missing `///` doc on public fields** (`favorites_section.dart` lines 12–16)
All five public fields (`calculations`, `estimations`, `onViewAll`, `onCalculationTap`, `onEstimationTap`) lack doc comments. Project conventions require `///` on all public members.

**4. Widget — class-level doc doesn't mention empty state or header row** (`favorites_section.dart` lines 9–10)
The two-line doc says "combining favorited calculations and cost estimations in a vertical list" but omits the empty state and the "View all" callback. A reader relying only on the doc comment would not know the full observable contract.

**5. Screenshot test — redundant `ensureInitialized()` call** (`favorites_section_screenshot_test.dart` line 11)
`TestWidgetsFlutterBinding.ensureInitialized()` is called standalone at the top of `main()`. `loadAppFontsAll()` already calls it internally, and `setUp` runs it again. The top-level call is harmless but noisy — remove it to match the peer screenshot test pattern.

**6. Separator index arithmetic** (`favorites_section.dart` lines 69 & 74) — *optional / lowest priority*
The `i < calculations.length - 1 || estimations.isNotEmpty` condition is correct (all four cases verified), but it's more complex than the Rule 12 spirit intended. Consider extracting a small `_withSeparators(List<Widget>)` helper so the loop body reads like prose. Non-blocking — correctness is fine as-is.

---

Full HTML report: ~/Desktop/pr-reviews/pr-review-construculator-app-390.html
